File: problemas/vmps-2/c++/flota.cc
#include "flota.h"

void Flota::annade(VMP* v)
{
	vmps.push_back(v);
}

const Flota::VMPS& Flota::lista() const
{
	return vmps;
}

VMP* Flota::masLejano() const
{
	VMP* tgt=nullptr;
	double maxd = 0;
	for (auto v : vmps)
	{
		double vd = v->distancia();
		if (vd>maxd)
		{
			tgt  = v;
			maxd = vd;
		}
	}

	return tgt;
}

Flota::VMPS Flota::fueraDeArea(double r) const
{
	VMPS out;

	for (auto v : vmps)
	{
		if (v->disponible() && !v->aparcado() && (v->distancia()>r))
		{
			v->bloquear();
			out.push_back(v);
		}
	}

	return out;
}

Flota::VMPS Flota::cargaBaja(double cargaMin) const
{
	VMPS out;

	for (auto v : vmps)
	{
		VMPElectrico* ve = dynamic_cast<VMPElectrico*>(v);
		if (ve)
		{
			if (!ve->cargando() && (ve->carga()<cargaMin))
			{
				ve->bloquear();
				out.push_back(ve);
			}
		}
	}

	return out;
}

double Flota::patinetesEnUso() const
{
	int   nu=0, np=0;
	double ratio = 0.0;

	for (auto v : vmps)
	{
		if (!v->disponible() && !v->aparcado())
		{
			nu++;
			if (dynamic_cast<Patinete*>(v)!=nullptr) np++;
		}
	}
	if (nu>0) ratio = 100*(double(np)/double(nu));

	return ratio;
}

File: problemas/vmps-2/java/Flota.java
import java.util.List;
import java.util.ArrayList;
import java.util.LinkedList;

class Flota
{
	private List<VMP> vmps;

	public Flota()
	{
		vmps = new ArrayList<VMP>();
		//vmps = new LinkedList<VMP>();
	}

	public void annade(VMP v)
	{
		vmps.add(v);
	}

	public List<VMP> lista()
	{
		return vmps;
	}

	public VMP masLejano()
	{
		VMP tgt=null;
		double maxd = 0;
		for (VMP v : vmps)
		{
			double vd = v.distancia();
			if (vd>maxd)
			{
				tgt  = v;
				maxd = vd;
			}
		}

		return tgt;
	}

	public List<VMP> fueraDeArea(double r)
	{
		List<VMP> out = new ArrayList<>();

		for (VMP v : vmps)
		{
			if (v.disponible() && !v.aparcado() && (v.distancia()>r))
			{
				v.bloquear();
				out.add(v);
			}
		}

		return out;
	}

	public List<VMP> cargaBaja(double cargaMin)
	{
		List<VMP> out = new ArrayList<>();

		for (VMP v : vmps)
		{
			if (v instanceof VMPElectrico)
			{
				VMPElectrico ve = (VMPElectrico)v;
				if ((!ve.cargando()) && (ve.carga()<cargaMin))
				{
					v.bloquear();
					out.add(v);
				}
			}
		}

		return out;
	}

	public double patinetesEnUso()
	{
		int   nu=0, np=0;
		double ratio = 0.0f;

		for (VMP v : vmps)
		{
			if (!v.disponible() && !v.aparcado())
			{
				nu++;
				if (v instanceof Patinete) np++;
			}
		}
		if (nu>0) ratio = 100*((double)np/(double)nu);

		return ratio;
	}
}

